core/sr-primitives/src/traits.rs

```diff
@@ -124,6 +124,49 @@ pub trait BlockNumberToHash {
 	}
 }
 
+/// Outcome of a balance update.
+pub enum UpdateBalanceOutcome {
+	/// Account balance was simply updated.
+	Updated,
+	/// The update has led to killing of the account.
+	AccountKilled,
+}
+
+pub trait TransferToken<AccountId> {
```

Comment:
I think we should keep this trait as simple as possible. Many methods here are internal/private methods.
This is my take: https://github.com/paritytech/substrate/pull/1648/files#diff-cfb20bdf2f3264df022885d7fe2de6a7R147

We should agree on a name to reduce upcoming conflicts...
`TransferToken` or `TransferAsset`?
`Balance` or `Amount`?

Reply:
Thanks, I have very few knowledge on blockchain names. I would be a bit more in favor of Balance than Amount as it characterize the type. But I won't argue at all, frankly I'm following your choice on that.

Yes the trait is bad in the current impl

Also if it happens you need the decl_event! modification I could make it in another PR

Reply:
what kind of internal private methods were you actually thinking of ?

Reply:
I don't see why other modules will want to call `increase_total_stake_by` (that's increase total issuance, nothing to do with staking).
Why anyone other than balances module needs to be aware of the different between `set_free_balance_creating` and `set_free_balance`?
I see many methods needs to be exposed only because contract module needs to reimplements the same logic with its overlay account db, if we can improve and avoid this duplication some how (which I don't have a good plan), then all the read config methods should be private (e.g. `creation_fee` and `transfer_fee`)